core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -631,12 +652,14 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
         case Some(currentReplicas) => currentReplicas == reassignmentReplicas // reassignment completed
       }
     }.keys
-    reassignedPartitions.foreach(removePartitionFromReassignedPartitions)
     val partitionsToReassign = partitionsBeingReassigned -- reassignedPartitions
+    // Must populate `partitionsBeingReassigned` before invoking `removePartitionFromReassignedPartitions` (see method
+    // documentation for the explanation)
```

Comment:
This is not really introduced in this patch. However, it seems that the code btw line 648 and 655 for checking whether the reassignmentReplicas are the same as the assigned replicas is the same as in maybeTriggerPartitionReassignment(topicPartitions(). The latter will be called after initializePartitionReassignment(). Perhaps we could just remove the duplicated logic here?

Reply:
It's a good point about reducing duplication. One downside is that the logging would be less clear in the case of a Controller restart while reassignment is in progress. With the current code, we get a nice summary:

```scala
info(s"Partitions being reassigned: $partitionsBeingReassigned")
info(s"Partitions already reassigned: $reassignedPartitions")
info(s"Resuming reassignment of partitions: $partitionsToReassign")
```

If we do the suggested change, then the second line becomes "Partitions where topics have been deleted" and the partitions that have completed the assignment would be logged later as:

```scala
info(s"Partition $tp to be reassigned is already assigned to replicas " +
                s"${newReplicas.mkString(",")}. Ignoring request for partition reassignment.")
```

We could add a parameter to `maybeTriggerPartitionReassignment` to try and improve the log message, but not sure the extra complexity is worth it. On the other hand, it's probably a bit rare that the reassignment completes and the Controller is restarted before the `reassign_partitions` znode is updated. Still, these cases are more likely to need investigation.

Given the above, do you think we should change it or keep as it is?

Reply:
Good question on logging. How about in initializePartitionReassignment(), we just log  partitionsBeingReassigned read from ZK? Other logging will just be done in maybeTriggerPartitionReassignment(). This way, the logging is actually more consistent btw the initial handling of partition reassignment and the controller failover.

Reply:
Pushed.